### ISSUE TITLE
require is not defined

### ISSUE THREAD

**Issue by Ptival** [facebook/flow#407]
Hello,

From some snippets here and there, I have figured out that I can use require('filename'); to import Flow modules.
I use the in-browser JSX transformer, because I don't quite understand how to set things up properly.

Now, I get this error:

````Uncaught ReferenceError: require is not defined````

It is unclear to me where require is supposed to come from. I assumed it could be that I need to use require.js to actually implement require, but that does not seem to work either:

````Uncaught Error: Mismatched anonymous define() module: function (){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(_dereq_,module,exports){````

I suppose I'm doing things very wrong but I am new to all of this and the documentation of flow is too vague on the subject.

**Comment by popham** [comment]
The in-browser transformer is for React components, not for Flow code (I'm pretty sure--I think the in-browser transformer is a low priority item:  `jsx --watch`-like workflows seem like the path of least resistence).

The require you've been using comes from `lib/core.js`:

```
declare var module: {
    require(id: string): any;
    // ...
};
declare function require(id: string): any;
```

This is Node's synchronous `require`.  RequireJS implements an asynchronous `require`, and the signature of its `require` contradicts the signature of Node's `require`.

Here's the quickest path to getting something in a browser that I can think of:

* Use [JSX precompiler](https://facebook.github.io/react/docs/tooling-integration.html#productionizing-precompiled-jsx) to strip type annotations (jsx --strip-types), and then
* Use [Browserify](http://browserify.org/) get yourself a compatible `require` (it bundles everything into a blob so that synchronous `require`s won't block while waiting on I/O).

For quick and dirty stuff, I'll get a `jsx ... && browserify ...` in my terminal history for reuse.  I guess under Flow this would become  `flow ... && jsx ... && browserify ...`.

**Comment by popham** [comment]
As for the documentation, there are so many different feasible workflows, I can't imagine blessing just one of them.

I think that the choice to lump Node signatures together with browser signatures in Flow's default library, e.g. WebSocket and Buffer, is sensible because of tools like *Browserify* and *webpack*.  I'm betting that this choice is the source of your confusion, though.  Lumping signatures from the two environments together implicitly blesses Browserify-like workflows.

Would some documentation on the decision to expose the two environments have avoided your confusion?

**Comment by Ptival** [comment]
@popham Thanks, your pointers help me set up a simple but reasonable workflow.

The main issue is that I am new to the JS ecosystem, so I did not even know about Node, Browserify, Webpack. So I figured out that I needed some library to implement this require, but did not know which one would fit the bill.

I now followed your approach of jsx --strip-types all my files, then browserify-ing the entry file.
